### PR TITLE
Add the presto router

### PR DESCRIPTION
**Pull request by ChunxuTang**
This PR adds a presto router service in front of presto clusters, helping routing requests to presto clusters. The router makes it easier to scale out presto clusters with an aggregated Presto-styled UI and APIs.

Co-Authored-By: Hao Luo <hluo@twitter.com>
Co-Authored-By: Beinan Wang <beinanw@twitter.com>

Test plan - Added tests

```
== RELEASE NOTES ==

General Changes
* Add a router in front of presto clusters. The router is deployed as a standalone service with a presto-styled UI. 

```

### CODE REVIEW

**Comment by ChunxuTang** [comment]
An example of the aggregated UI is shown below. It is served by the presto router.

router-ui

**Comment by ChunxuTang** [comment]
Hi @zhenxiao and @beinan, would you mind reviewing this new module? Thanks!

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        if (!target.isPresent()) {
+            return Optional.empty();
+        }
+
+        checkArgument(groups.containsKey(target.get()));
+        List<URI> candidates = groups.get(target.get()).getMembers();
+        return Optional.ofNullable(candidates.get(RANDOM.nextInt(candidates.size())));
+    }
+
+    public Optional<String> matchGroup(RequestInfo requestInfo)
```
Comment (rongrong) [inline presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java:109]: Does this need to be public?

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        this.clusterManager = requireNonNull(clusterManager, "clusterManager is null");
+        this.remoteInfoFactory = requireNonNull(remoteInfoFactory, "remoteInfoFactory is null");
+        this.queryInfoUpdateExecutor = newSingleThreadScheduledExecutor(threadsNamed("query-info-poller-%s"));
+    }
+
+    @PostConstruct
+    public void startPollingQueryInfo()
+    {
+        clusterManager.getAllClusters().stream()
```
Comment (rongrong) [inline presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java:63]: nits: `stream()` is probably not needed

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                log.error(e, "Error polling list of queries");
+            }
+        }, 5, 5, TimeUnit.SECONDS);
+
+        pollQueryInfos();
+    }
+
+    private void pollQueryInfos()
+    {
+        Set<URI> allClusters = newHashSet(clusterManager.getAllClusters());
```
Comment (rongrong) [inline presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java:83]: nits: Prefer `ImmutableSet`

**Review by rongrong — APPROVED** [review]
Thanks a lot for contributing the project! Just some nits.

**Comment by rongrong** [comment]
Please rebase and fix the merge conflict.

**Review by 7c00 — APPROVED** [review]
LGTM

File: presto-router/src/main/java/com/facebook/presto/router/RouterModule.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import static com.facebook.airlift.http.server.HttpServerBinder.httpServerBinder;
+import static com.facebook.airlift.jaxrs.JaxrsBinder.jaxrsBinder;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class RouterModule
+        extends AbstractConfigurationAwareModule
+{
+    private static final int IDLE_TIMEOUT_SECOND = 30;
+    private static final int REQUEST_TIMEOUT_SECOND = 10;
+    private static final String QUERY_TRACKER = "query-tracker";
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/RouterModule.java:41]: Thanks for the suggestion! I've made the style here consistent.

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java (hunk 1 above)
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java]: Thanks for pointing that out! I updated it to `private`.

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java (hunk 2 above)
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java]: Yep. I removed the unnecessary `stream()` from the code (there's another place I used it redundantly and I also removed it from there).

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java (hunk 3 above)
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java]: Nice catch! I updated the usage of `Set` to `ImmutableSet`.

**Comment by ChunxuTang** [comment]
> Please extend the README with proper documentation on how to use it. Also consider add to official documentation.

Thanks a lot for the review!

I've rebased and fixed some merge conflicts. I've also updated the README file with necessary tutorials about how to use the module and the UI. Plan to send another PR to update the overall Presto doc for the new module.

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+import static com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toMap;
+
+public class ClusterManager
+{
+    private static final Random RANDOM = new Random();
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java:48]: Generally, there's only one cluster management instance as we usually don't have many Presto clusters running simultaneously (e.g., only ~10 at most). So we don't have to introduce thread-local variables for the current use cases. But I do agree that in the future when we need to handle a much higher number of Presto clusters, we'll need to introduce some concurrency mechanisms.

File: presto-router/README.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+Presto Router
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-router/README.txt]: Thanks for the suggestion! I've added the instructions to launch this service in IntelliJ.

File: presto-router/src/main/java/com/facebook/presto/router/cluster/RemoteInfoFactory.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.net.URI;
+
+import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static java.util.Objects.requireNonNull;
+
+public class RemoteInfoFactory
+{
+    private static final String QUERY_INFO = "/v1/query";
+    private static final String CLUSTER_INFO = "/v1/cluster";
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/RemoteInfoFactory.java:28]: That'll make it clearer! Thanks for the suggestion. I've updated the code.

File: presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        clusterManager.getAllClusters().stream()
+                .forEach(uri -> {
+                    remoteClusterInfos.put(uri, remoteInfoFactory.createRemoteClusterInfo(uri));
+                    remoteQueryInfos.put(uri, remoteInfoFactory.createRemoteQueryInfo(uri));
+                });
+
+        queryInfoUpdateExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                pollQueryInfos();
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java:69]: As the router usually doesn't need to deal with a large num of Presto clusters (e.g. only ~10 at most), scheduling with a fixed delay generally looks sufficient. Maybe we can add a config to let users to customize the delay/jitter here.

File: presto-router/src/main/java/com/facebook/presto/router/cluster/RequestInfo.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        return clientTags;
+    }
+
+    private List<String> parseClientTags(HttpServletRequest servletRequest)
+    {
+        return ImmutableList.copyOf(SPLITTER.split(nullToEmpty(servletRequest.getHeader(PRESTO_CLIENT_TAGS))));
+    }
+
+    private String parseHeader(HttpServletRequest servletRequest, String header)
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/RequestInfo.java]: Nice catch! Updated the code.

File: presto-router/src/main/java/com/facebook/presto/router/cluster/RequestInfo.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        return source;
+    }
+
+    public List<String> getClientTags()
+    {
+        return clientTags;
+    }
+
+    private List<String> parseClientTags(HttpServletRequest servletRequest)
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-router/src/main/java/com/facebook/presto/router/cluster/RequestInfo.java]: Thanks! Updated the code.

**Comment by ChunxuTang** [comment]
@7c00 Thanks a lot for your help!

**Comment by arunthirupathi** [comment]
The tests are flaky

https://github.com/prestodb/presto/runs/7398294709?check_suite_focus=true#step:7:21636

Error:  com.facebook.presto.router.TestClusterManager.testQuery  Time elapsed: 20.133 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:110)

2022-07-18T16:51:51.463-0500	INFO	TestNG-test=Surefire test-1	com.facebook.airlift.bootstrap.LifeCycleManager	Life cycle stopped.
	at com.facebook.presto.router.TestClusterManager.assertQueryState(TestClusterManager.java:149)
	at com.facebook.presto.router.TestClusterManager.testQuery(TestClusterManager.java:129)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)

**Comment by ChunxuTang** [comment]
@arunthirupathi Thanks for figuring that out!
I've created https://github.com/prestodb/presto/pull/18063 to fix that. Would you mind having a quick review of it?

**Comment by ChunxuTang** [comment]
FYI: The usage of the presto router is documented in https://github.com/prestodb/presto/pull/18118

**Comment by highker** [comment]
Another flaky test fix: #18133

**Pull request merged**